allennlp/training/trainer.py

```diff
@@ -1105,9 +1014,8 @@ def _try_train(self) -> Dict[str, Any]:
                 logger.info("Estimated training time remaining: %s", formatted_time)
 
             epochs_trained += 1
-
-        for callback in self._end_callbacks:
-            callback(self, metrics=metrics, epoch=epoch, is_primary=self._primary)
+        else:
+            epoch = self._num_epochs - 1
```

Comment:
`for ... else` ... a real rarity. This fixes the situation where the training had already finished, was checkpointed, but something failed during validation/test, and you want to run again, I assume? Looks quite reasonable, but I wanted to make sure it's on purpose.

Reply:
Consider you want to `--recover` the training, but it's already done, so you basically run `for epoch in range(epoch_counter, self._num_epochs):` where `epoch_counter == self._num_epochs`. This means that the loop will not run at all.
The `epoch` parameter, in that case, will be unbound.

Minimal example will be:
```python
In [1]: for i in range(10, 10):
   ...:     pass
   ...:
   ...: print(i)
```
This will output:
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-58b7096fed5d> in <module>
      2     pass
      3
----> 4 print(i)

NameError: name 'i' is not defined
```

Reply:
Yep, I get it. I just wanted to make sure it's intentional, because most people don't even know that `for ... else` exists :-)